src/Nethermind/Nethermind.Era1/EraImporter.cs

```diff
@@ -0,0 +1,289 @@
+        if (to == 0) to = long.MaxValue;
+        if (to != long.MaxValue && lastBlockInStore < to)
+        {
+            throw new EraImportException($"The directory given for import '{src}' have highest block number {lastBlockInStore} which is lower then last requested block {to}.");
+        }
+        if (to == long.MaxValue)
+        {
+            to = lastBlockInStore;
+        }
+
+        long firstBlockInStore = eraStore.FirstBlock;
+        if (from == 0 && firstBlockInStore != 0)
+        {
+            from = firstBlockInStore;
+        }
+        else if (from < firstBlockInStore)
+        {
+            throw new EraImportException($"The directory given for import '{src}' have lowest block number {firstBlockInStore} which is lower then last requested block {from}.");
+        }
+        if (from > to && to != 0)
+            throw new ArgumentException("Start block must not be after end block", nameof(from));
+
+        _receiptsDb.Tune(ITunableDb.TuneType.HeavyWrite);
+        _blocksDb.Tune(ITunableDb.TuneType.HeavyWrite);
+
+        return ImportInternal(from, to, eraStore, cancellation)
+            .ContinueWith((t) =>
+            {
+                _receiptsDb.Tune(ITunableDb.TuneType.Default);
+                _blocksDb.Tune(ITunableDb.TuneType.Default);
+
+                if (t.IsFaulted)
+                    throw t.Exception?.InnerException!;
+            }, cancellation);
+    }
+
+    private async Task ImportInternal(
+        long from,
+        long to,
+        IEraStore eraStore,
+        CancellationToken cancellation)
+    {
+        if (_logger.IsInfo) _logger.Info($"Starting history import from {from} to {to}");
+
+        using IEraStore _ = eraStore;
+
+        DateTime lastProgress = DateTime.Now;
+        DateTime startTime = DateTime.Now;
+        TimeSpan elapsed = TimeSpan.Zero;
+        long totalblocks = to - from + 1;
+        long blocksProcessed = 0;
+        long blocksProcessedAtLastLog = 0;
+
+        using BlockTreeSuggestPacer pacer = new BlockTreeSuggestPacer(_blockTree);
+        long blockNumber = from;
+
+        long suggestFromBlock = (_blockTree.Head?.Number ?? 0) + 1;
+        if (_syncConfig.FastSync && suggestFromBlock == 1)
+        {
+            // Its syncing right now. So no state.
+            suggestFromBlock = long.MaxValue;
+        }
+
+        // Add last header first.
+        // This set BestSuggestedHeader so that the receipt insert does not create tx index unnecessarily
+        {
+            (Block? b, TxReceipt[]? _) = (await eraStore.FindBlockAndReceipts(eraStore.LastBlock, cancellation: cancellation))!;
+            _blockTree.Insert(b!.Header);
+        }
+
+        // I wish I could say that EraStore can be run used in parallel in any way you like but I could not make it so.
+        // This make the `blockNumber` aligned to era file boundary so that when running parallel, each thread does not
+        // work on the same era file as other thread.
```

Comment:
The EraReader sharing in EraStore is limited. If a thread try to access an era reader that is currently being used, it will create a new EraReader. Need a more sophisticated reference counting mechanism to make work.